# The Power of Electric Snakes!
## An Introduction to PySpark

##Section 0: Initial Data Preparation

In [3]:
from pyspark.sql import *

lines = sc.textFile("/databricks-datasets/adult/adult.data")
adult_rdd = lines \
  .filter(lambda f: f != '') \
  .map(lambda l: tuple(l.split(', '))) \
  .map(lambda r: ( \
    int(r[0]) \
    ,r[1] \
    ,int(r[2]) \
    ,r[3] \
    ,int(r[4]) \
    ,r[5] \
    ,r[6] \
    ,r[7] \
    ,r[8] \
    ,r[9] \
    ,int(r[10]) \
    ,int(r[11]) \
    ,int(r[12]) \
    ,r[13] \
    ,r[14] \
  ))
adult_df = adult_rdd \
  .map(lambda r: Row( \
    age = r[0] \
    ,workclass = r[1] \
    ,fnlwgt = r[2] \
    ,education = r[3] \
    ,educationnum = r[4] \
    ,maritalstatus = r[5] \
    ,occupation = r[6] \
    ,relationship = r[7] \
    ,race = r[8] \
    ,sex = r[9] \
    ,capitalgain = r[10] \
    ,capitalloss = r[11] \
    ,hoursperweek = r[12] \
    ,nativecountry = r[13] \
    ,income = r[14]) \
  ) \
  .toDF()

## Section 1: RDDs vs. DataFrames

Generally, RDD's are viewed by using .collect() and DataFrames are viewed using .show()

In [6]:
adult_rdd.collect()

Out[2]: [(39,
 'State-gov',
 77516,
 'Bachelors',
 13,
 'Never-married',
 'Adm-clerical',
 'Not-in-family',
 'White',
 'Male',
 2174,
 0,
 40,
 'United-States',
 '<=50K'),
 (50,
 'Self-emp-not-inc',
 83311,
 'Bachelors',
 13,
 'Married-civ-spouse',
 'Exec-managerial',
 'Husband',
 'White',
 'Male',
 0,
 0,
 13,
 'United-States',
 '<=50K'),
 (38,
 'Private',
 215646,
 'HS-grad',
 9,
 'Divorced',
 'Handlers-cleaners',
 'Not-in-family',
 'White',
 'Male',
 0,
 0,
 40,
 'United-States',
 '<=50K'),
 (53,
 'Private',
 234721,
 '11th',
 7,
 'Married-civ-spouse',
 'Handlers-cleaners',
 'Husband',
 'Black',
 'Male',
 0,
 0,
 40,
 'United-States',
 '<=50K'),
 (28,
 'Private',
 338409,
 'Bachelors',
 13,
 'Married-civ-spouse',
 'Prof-specialty',
 'Wife',
 'Black',
 'Female',
 0,
 0,
 40,
 'Cuba',
 '<=50K'),
 (37,
 'Private',
 284582,
 'Masters',
 14,
 'Married-civ-spouse',
 'Exec-managerial',
 'Wife',
 'White',
 'Female',
 0,
 0,
 40,
 'United-States',
 '<=50K'),
 (49,
 'Private',
 160187,
 '9th',
 5,
 'Married-spouse-absent',
 'Other-service',
 'Not-in-family',
 'Black',
 'Female',
 0,
 0,
 16,
 'Jamaica',
 '<=50K'),
 (52,
 'Self-emp-not-inc',
 209642,
 'HS-grad',
 9,
 'Married-civ-spouse',
 'Exec-managerial',
 'Husband',
 'White',
 'Male',
 0,
 0,
 45,
 'United-States',
 '>50K'),
 (31,
 'Private',
 45781,
 'Masters',
 14,
 'Never-married',
 'Prof-specialty',
 'Not-in-family',
 'White',
 'Female',
 14084,
 0,
 50,
 'United-States',
 '>50K'),
 (42,
 'Private',
 159449,
 'Bachelors',
 13,
 'Married-civ-spouse',
 'Exec-managerial',
 'Husband',
 'White',
 'Male',
 5178,
 0,
 40,
 'United-States',
 '>50K'),
 (37,
 'Private',
 280464,
 'Some-college',
 10,
 'Married-civ-spouse',
 'Exec-managerial',
 'Husband',
 'Black',
 'Male',
 0,
 0,
 80,
 'United-States',
 '>50K'),
 (30,
 'State-gov',
 141297,
 'Bachelors',
 13,
 'Married-civ-spouse',
 'Prof-specialty',
 'Husband',
 'Asian-Pac-Islander',
 'Male',
 0,
 0,
 40,
 'India',
 '>50K'),
 (23,
 'Private',
 122272,
 'Bachelors',
 13,
 'Never-married',
 'Adm-clerical',
 'Own-child',
 'White',
 'Female',
 0,
 0,
 30,
 'United-States',
 '<=50K'),
 (32,
 'Private',
 205019,
 'Assoc-acdm',
 12,
 'Never-married',
 'Sales',
 'Not-in-family',
 'Black',
 'Male',
 0,
 0,
 50,
 'United-States',
 '<=50K'),
 (40,
 'Private',
 121772,
 'Assoc-voc',
 11,
 'Married-civ-spouse',
 'Craft-repair',
 'Husband',
 'Asian-Pac-Islander',
 'Male',
 0,
 0,
 40,
 '?',
 '>50K'),
 (34,
 'Private',
 245487,
 '7th-8th',
 4,
 'Married-civ-spouse',
 'Transport-moving',
 'Husband',
 'Amer-Indian-Eskimo',
 'Male',
 0,
 0,
 45,
 'Mexico',
 '<=50K'),
 (25,
 'Self-emp-not-inc',
 176756,
 'HS-grad',
 9,
 'Never-married',
 'Farming-fishing',
 'Own-child',
 'White',
 'Male',
 0,
 0,
 35,
 'United-States',
 '<=50K'),
 (32,
 'Private',
 186824,
 'HS-grad',
 9,
 'Never-married',
 'Machine-op-inspct',
 'Unmarried',
 'White',
 'Male',
 0,
 0,
 40,
 'United-States',
 '<=50K'),
 (38,
 'Private',
 28887,
 '11th',
 7,
 'Married-civ-spouse',
 'Sales',
 'Husband',
 'White',
 'Male',
 0,
 0,
 50,
 'United-States',
 '<=50K'),
 (43,
 'Self-emp-not-inc',
 292175,
 'Masters',
 14,
 'Divorced',
 'Exec-managerial',
 'Unmarried',
 'White',
 'Female',
 0,
 0,
 45,
 'United-States',
 '>50K'),
 (40,
 'Private',
 193524,
 'Doctorate',
 16,
 'Married-civ-spouse',
 'Prof-specialty',
 'Husband',
 'White',
 'Male',
 0,
 0,
 60,
 'United-States',
 '>50K'),
 (54,
 'Private',
 302146,
 'HS-grad',
 9,
 'Separated',
 'Other-service',
 'Unmarried',
 'Black',
 'Female',
 0,
 0,
 20,
 'United-States',
 '<=50K'),
 (35,
 'Federal-gov',
 76845,
 '9th',
 5,
 'Married-civ-spouse',
 'Farming-fishing',
 'Husband',
 'Black',
 'Male',
 0,
 0,
 40,
 'United-States',
 '<=50K'),
 (43,
 'Private',
 117037,
 '11th',
 7,
 'Married-civ-spouse',
 'Transport-moving',
 'Husband',
 'White',
 'Male',
 0,
 2042,
 40,
 'United-States',
 '<=50K'),
 (59,
 'Private',
 109015,
 'HS-grad',
 9,
 'Divorced',
 'Tech-support',
 'Unmarried',
 'White',
 'Female',
 0,
 0,
 40,
 'United-States',
 '<=50K'),
 (

In [7]:
adult_df.show()

+---+-----------+-----------+------------+------------+------+------------+------+--------------------+-------------+-----------------+------------------+-------------+------+----------------+
age|capitalgain|capitalloss| education|educationnum|fnlwgt|hoursperweek|income| maritalstatus|nativecountry| occupation| race| relationship| sex| workclass|
+---+-----------+-----------+------------+------------+------+------------+------+--------------------+-------------+-----------------+------------------+-------------+------+----------------+
 39| 2174| 0| Bachelors| 13| 77516| 40| <=50K| Never-married|United-States| Adm-clerical| White|Not-in-family| Male| State-gov|
 50| 0| 0| Bachelors| 13| 83311| 13| <=50K| Married-civ-spouse|United-States| Exec-managerial| White| Husband| Male|Self-emp-not-inc|
 38| 0| 0| HS-grad| 9|215646| 40| <=50K| Divorced|United-States|Handlers-cleaners| White|Not-in-family| Male| Private|
 53| 0| 0| 11th| 7|234721| 40| <=50K| Married-civ-spouse|United-States|Handlers-cleaners| Black| Husband| Male| Private|
 28| 0| 0| Bachelors| 13|338409| 40| <=50K| Married-civ-spouse| Cuba| Prof-specialty| Black| Wife|Female| Private|
 37| 0| 0| Masters| 14|284582| 40| <=50K| Married-civ-spouse|United-States| Exec-managerial| White| Wife|Female| Private|
 49| 0| 0| 9th| 5|160187| 16| <=50K|Married-spouse-ab...| Jamaica| Other-service| Black|Not-in-family|Female| Private|
 52| 0| 0| HS-grad| 9|209642| 45| >50K| Married-civ-spouse|United-States| Exec-managerial| White| Husband| Male|Self-emp-not-inc|
 31| 14084| 0| Masters| 14| 45781| 50| >50K| Never-married|United-States| Prof-specialty| White|Not-in-family|Female| Private|
 42| 5178| 0| Bachelors| 13|159449| 40| >50K| Married-civ-spouse|United-States| Exec-managerial| White| Husband| Male| Private|
 37| 0| 0|Some-college| 10|280464| 80| >50K| Married-civ-spouse|United-States| Exec-managerial| Black| Husband| Male| Private|
 30| 0| 0| Bachelors| 13|141297| 40| >50K| Married-civ-spouse| India| Prof-specialty|Asian-Pac-Islander| Husband| Male| State-gov|
 23| 0| 0| Bachelors| 13|122272| 30| <=50K| Never-married|United-States| Adm-clerical| White| Own-child|Female| Private|
 32| 0| 0| Assoc-acdm| 12|205019| 50| <=50K| Never-married|United-States| Sales| Black|Not-in-family| Male| Private|
 40| 0| 0| Assoc-voc| 11|121772| 40| >50K| Married-civ-spouse| ?| Craft-repair|Asian-Pac-Islander| Husband| Male| Private|
 34| 0| 0| 7th-8th| 4|245487| 45| <=50K| Married-civ-spouse| Mexico| Transport-moving|Amer-Indian-Eskimo| Husband| Male| Private|
 25| 0| 0| HS-grad| 9|176756| 35| <=50K| Never-married|United-States| Farming-fishing| White| Own-child| Male|Self-emp-not-inc|
 32| 0| 0| HS-grad| 9|186824| 40| <=50K| Never-married|United-States|Machine-op-inspct| White| Unmarried| Male| Private|
 38| 0| 0| 11th| 7| 28887| 50| <=50K| Married-civ-spouse|United-States| Sales| White| Husband| Male| Private|
 43| 0| 0| Masters| 14|292175| 45| >50K| Divorced|United-States| Exec-managerial| White| Unmarried|Female|Self-emp-not-inc|
+---+-----------+-----------+------------+------------+------+------------+------+--------------------+-------------+-----------------+------------------+-------------+------+----------------+
only showing top 20 rows

We see that the RDD is represented as an unnamed tuple of objects.  Tuples can also other Tuples, allowing us to create RDDs containing tuples of tuples of tuples, creating a multidimensional data type.  DataFrames are what we would typically think of as tables, two-dimensional with fixed data types for each column.

Data Transformations are really where RDDs and DataFrames deviate.  Let's start with a simple example, creating a new RDD/DataFrame that contains "age" and "hourspermonth" = hoursperweek * 4.

In [10]:
adult_rdd_t = adult_rdd \
  .map(lambda r: (r[0], r[12] * 4))

adult_rdd_t.top(5)

Out[4]: [(90, 396), (90, 288), (90, 240), (90, 220), (90, 200)]

In [11]:
from pyspark.sql.functions import col

adult_df_t = adult_df \
  .withColumn('hourspermonth', col('hoursperweek') * 4) \
  .select('age', 'hourspermonth')

adult_df_t.show(5)

+---+-------------+
age|hourspermonth|
+---+-------------+
 39| 160|
 50| 52|
 38| 160|
 53| 160|
 28| 160|
+---+-------------+
only showing top 5 rows

We can see that RDDs are manipulated using the .map() function, with the fields denoted by the ordinal location in the tuple.  DataFrames on the other hand are treated more like tables, with the ability to create columns with .withColumn() and select columns using .select().  DataFrames also give us the ability to reference columns by name using the col() function.

Next, we'll make it even more complex with aggregations.  Let's try simply calculating the number of rows the RDD and DataFrame.

In [14]:
adult_rdd_c = adult_rdd \
  .count()

adult_rdd_c

Out[6]: 32561

In [15]:
adult_df_c = adult_df \
  .count()

adult_df_c

Out[7]: 32561

As it turns out, RDDs and DataFrames can both leverage the .count() function.

Now, let's try to calculate the sum of the "hoursperweek" field.

In [18]:
adult_rdd_s = adult_rdd \
  .map(lambda r: r[12]) \
  .reduce(lambda t, h: t + h)

adult_rdd_s

Out[8]: 1316684

In [19]:
adult_df_s = adult_df \
  .groupBy() \
  .sum('hoursperweek')

adult_df_s.show()

+-----------------+
sum(hoursperweek)|
+-----------------+
 1316684|
+-----------------+

Now that we're doing more complex calculations, the differences between RDDs and DataFrames become apparent again.  RDDs, along with the .map() function, also leverage a .reduce() function.  The syntax for this function is not intuitive to most data professionals and can be cumbersome.  DataFrames leverage familiar SQL concepts, like .groupBy() and .sum(), making them much more approachable.

Now, let's move on to something even more complex.  Let's calculate the average age for each level of "education".

In [22]:
adult_rdd_a = adult_rdd \
  .map(lambda r: (r[3], r[0])) \
  .aggregateByKey( \
    (0,0) \
    ,lambda w, r: (w[0] + r, w[1] + 1) \
    ,lambda t, p: (t[0] + p[0], t[1] + p[1]) \
  )

adult_rdd_a.top(5)

Out[10]: [('Some-college', (260699, 7291)),
 ('Prof-school', (25774, 576)),
 ('Preschool', (2181, 51)),
 ('Masters', (75898, 1723)),
 ('HS-grad', (409271, 10501))]

In [23]:
adult_rdd_a = adult_rdd \
  .map(lambda r: (r[3], r[0])) \
  .aggregateByKey( \
    (0,0) \
    ,lambda w, r: (w[0] + r, w[1] + 1) \
    ,lambda t, p: (t[0] + p[0], t[1] + p[1]) \
  ) \
  .map(lambda r: (r[0], r[1][0] / r[1][1]))

adult_rdd_a.top(5)

Out[11]: [('Some-college', 35.75627485941572),
 ('Prof-school', 44.74652777777778),
 ('Preschool', 42.76470588235294),
 ('Masters', 44.04991294254208),
 ('HS-grad', 38.974478621083705)]

In [24]:
adult_df_a = adult_df \
  .groupBy('education') \
  .avg('age')

adult_df_a.show(5)

+----------+------------------+
 education| avg(age)|
+----------+------------------+
 Masters| 44.04991294254208|
 10th| 37.42979635584137|
 5th-6th|42.885885885885884|
Assoc-acdm|37.381443298969074|
 Assoc-voc| 38.55354558610709|
+----------+------------------+
only showing top 5 rows

At this point, RDDs are becoming incredible complex to manipulate.  The .aggregateByKey() requires defining Within-Partition and Across-Partition aggregation functions, resulting in extremely complex logic to perform simple transformations.  Again, DataFrames are much approachable to those with a SQL background, leveraging the .groupBy() and .avg() functions.

Finally, let's join the average ages back onto the original RDD and DataFrame.

In [27]:
adult_rdd_j = adult_rdd \
  .map(lambda r: ( \
    r[3] \
    ,( \
      r[0] \
      ,r[1] \
      ,r[2] \
      ,r[4] \
      ,r[5] \
      ,r[6] \
      ,r[7] \
      ,r[8] \
      ,r[9] \
      ,r[10] \
      ,r[11] \
      ,r[12] \
      ,r[13] \
      ,r[14] \
    ) \
  )) \
  .join(adult_rdd_a)

adult_rdd_j.top(5)

Out[13]: [('Some-college',
 ((90,
 'Self-emp-not-inc',
 282095,
 10,
 'Married-civ-spouse',
 'Farming-fishing',
 'Husband',
 'White',
 'Male',
 0,
 0,
 40,
 'United-States',
 '<=50K'),
 35.75627485941572)),
 ('Some-college',
 ((90,
 'Private',
 171956,
 10,
 'Separated',
 'Adm-clerical',
 'Own-child',
 'White',
 'Female',
 0,
 0,
 40,
 'Puerto-Rico',
 '<=50K'),
 35.75627485941572)),
 ('Some-college',
 ((90,
 'Private',
 139660,
 10,
 'Divorced',
 'Sales',
 'Unmarried',
 'Black',
 'Female',
 0,
 0,
 37,
 'United-States',
 '<=50K'),
 35.75627485941572)),
 ('Some-college',
 ((90,
 'Private',
 52386,
 10,
 'Never-married',
 'Other-service',
 'Not-in-family',
 'Asian-Pac-Islander',
 'Male',
 0,
 0,
 35,
 'United-States',
 '<=50K'),
 35.75627485941572)),
 ('Some-college',
 ((90,
 'Private',
 52386,
 10,
 'Never-married',
 'Other-service',
 'Not-in-family',
 'Asian-Pac-Islander',
 'Male',
 0,
 0,
 35,
 'United-States',
 '<=50K'),
 35.75627485941572))]

In [28]:
adult_rdd_j = adult_rdd \
  .map(lambda r: ( \
    r[3] \
    ,( \
      r[0] \
      ,r[1] \
      ,r[2] \
      ,r[4] \
      ,r[5] \
      ,r[6] \
      ,r[7] \
      ,r[8] \
      ,r[9] \
      ,r[10] \
      ,r[11] \
      ,r[12] \
      ,r[13] \
      ,r[14] \
    ) \
  )) \
  .join(adult_rdd_a) \
  .map(lambda r: ( \
    r[1][0][0] \
    ,r[1][0][1] \
    ,r[1][0][2] \
    ,r[0] \
    ,r[1][0][3] \
    ,r[1][0][4] \
    ,r[1][0][5] \
    ,r[1][0][6] \
    ,r[1][0][7] \
    ,r[1][0][8] \
    ,r[1][0][9] \
    ,r[1][0][10] \
    ,r[1][0][11] \
    ,r[1][0][12] \
    ,r[1][0][13] \
    ,r[1][1] \
  ))

adult_rdd_j.top(5)

Out[14]: [(90,
 'Self-emp-not-inc',
 282095,
 'Some-college',
 10,
 'Married-civ-spouse',
 'Farming-fishing',
 'Husband',
 'White',
 'Male',
 0,
 0,
 40,
 'United-States',
 '<=50K',
 35.75627485941572),
 (90,
 'Self-emp-not-inc',
 155981,
 'Bachelors',
 13,
 'Married-civ-spouse',
 'Prof-specialty',
 'Husband',
 'White',
 'Male',
 10566,
 0,
 50,
 'United-States',
 '<=50K',
 38.90494864612511),
 (90,
 'Self-emp-not-inc',
 82628,
 'HS-grad',
 9,
 'Never-married',
 'Exec-managerial',
 'Not-in-family',
 'White',
 'Male',
 2964,
 0,
 12,
 'United-States',
 '<=50K',
 38.974478621083705),
 (90,
 'Private',
 313986,
 '11th',
 7,
 'Never-married',
 'Handlers-cleaners',
 'Own-child',
 'White',
 'Male',
 0,
 0,
 40,
 'United-States',
 '<=50K',
 32.35574468085106),
 (90,
 'Private',
 313986,
 '11th',
 7,
 'Married-civ-spouse',
 'Craft-repair',
 'Husband',
 'White',
 'Male',
 0,
 0,
 40,
 'United-States',
 '<=50K',
 32.35574468085106)]

In [29]:
adult_df_j = adult_df \
  .join(adult_df_a, on='education')

adult_df_j.show(5)

+---------+---+-----------+-----------+------------+------+------------+------+------------------+-------------+-------------+------------------+-------------+------+---------+-----------------+
education|age|capitalgain|capitalloss|educationnum|fnlwgt|hoursperweek|income| maritalstatus|nativecountry| occupation| race| relationship| sex|workclass| avg(age)|
+---------+---+-----------+-----------+------------+------+------------+------+------------------+-------------+-------------+------------------+-------------+------+---------+-----------------+
 10th| 67| 0| 0| 6|212759| 2| <=50K|Married-civ-spouse|United-States| ?| White| Husband| Male| ?|37.42979635584137|
 10th| 17| 34095| 0| 6|304873| 32| <=50K| Never-married|United-States| ?| White| Own-child|Female| ?|37.42979635584137|
 10th| 25| 0| 0| 6|255004| 40| <=50K| Never-married|United-States| Craft-repair| White|Not-in-family| Male| Private|37.42979635584137|
 10th| 60| 0| 0| 6| 24215| 10| <=50K| Divorced|United-States| ?|Amer-Indian-Eskimo|Not-in-family|Female| ?|37.42979635584137|
 10th| 59| 0| 0| 6|171328| 30| <=50K| Widowed|United-States|Other-service| Black| Unmarried|Female|Local-gov|37.42979635584137|
+---------+---+-----------+-----------+------------+------+------------+------+------------------+-------------+-------------+------------------+-------------+------+---------+-----------------+
only showing top 5 rows

We see that joining RDDs requires creating Key-Value pairs, which requires some additional mapping.  DataFrames can be joined using the simpler .join(on=) syntax.

##Section 2: Data Science Preparation

The first step of any Data Science Analysis is to summarize the data.

In [33]:
adult_df \
  .describe() \
  .show()

+-------+------------------+------------------+------------------+------------+------------------+------------------+------------------+------+-------------+-------------+----------------+------------------+------------+------+-----------+
summary| age| capitalgain| capitalloss| education| educationnum| fnlwgt| hoursperweek|income|maritalstatus|nativecountry| occupation| race|relationship| sex| workclass|
+-------+------------------+------------------+------------------+------------+------------------+------------------+------------------+------+-------------+-------------+----------------+------------------+------------+------+-----------+
 count| 32561| 32561| 32561| 32561| 32561| 32561| 32561| 32561| 32561| 32561| 32561| 32561| 32561| 32561| 32561|
 mean| 38.58164675532078|1077.6488437087312| 87.303829734959| null| 10.0806793403151|189778.36651208502|40.437455852092995| null| null| null| null| null| null| null| null|
 stddev|13.640432553581343| 7385.292084840351|402.96021864899984| null|2.5727203320673846|105549.97769702232|12.347428681731833| null| null| null| null| null| null| null| null|
 min| 17| 0| 0| 10th| 1| 12285| 1| <=50K| Divorced| ?| ?|Amer-Indian-Eskimo| Husband|Female| ?|
 max| 90| 99999| 4356|Some-college| 16| 1484705| 99| >50K| Widowed| Yugoslavia|Transport-moving| White| Wife| Male|Without-pay|
+-------+------------------+------------------+------------------+------------+------------------+------------------+------------------+------+-------------+-------------+----------------+------------------+------------+------+-----------+

The .describe() function is a great way for us to see summary statistics of our DataFrame.  Since the count for each column equals the total count of records, we know there are no Null values.  However, this doesn't mean that there aren't any junk values in the columns.  We can see a little bit of this in the table above.  There are some troubling things here though.

<br\>

* Is the capitalgain of 99999 legitimate?
* Is the hoursperweek of 99 legitimate?
* What's in the text columns?

Let's start with the numeric ones.

In [35]:
c = 'capitalgain'
adult_df \
  .groupBy(c) \
  .count() \
  .sort(col(c).desc()) \
  .show()

+-----------+-----+
capitalgain|count|
+-----------+-----+
 99999| 159|
 41310| 2|
 34095| 5|
 27828| 34|
 25236| 11|
 25124| 4|
 22040| 1|
 20051| 37|
 18481| 2|
 15831| 6|
 15024| 347|
 15020| 5|
 14344| 26|
 14084| 41|
 13550| 27|
 11678| 2|
 10605| 12|
 10566| 6|
 10520| 43|
 9562| 4|
+-----------+-----+
only showing top 20 rows

In [36]:
c = 'hoursperweek'
adult_df \
  .groupBy(c) \
  .count() \
  .sort(col(c).desc()).show()

+------------+-----+
hoursperweek|count|
+------------+-----+
 99| 85|
 98| 11|
 97| 2|
 96| 5|
 95| 2|
 94| 1|
 92| 1|
 91| 3|
 90| 29|
 89| 2|
 88| 2|
 87| 1|
 86| 2|
 85| 13|
 84| 45|
 82| 1|
 81| 3|
 80| 133|
 78| 8|
 77| 6|
+------------+-----+
only showing top 20 rows

Looking at capitalgain, we see a tremendous jump between the highest value (99999) and the second highest value (41310).  The highest value also has a relatively large number of records, this leads us to believe that this is not a legitimate value.  hoursperweek on the other hand, has a seamless transition from 98 to 99.  99 is a little top-heavy, but this could easily be because they reported any value higher than 99 as being 99.  Therefore, we don't see strong evidence that this data is illegitimate.  Let's move on to the string values.

In [38]:
c = 'education'
adult_df \
  .groupBy(c) \
  .count() \
  .sort(col(c)).show()

+------------+-----+
 education|count|
+------------+-----+
 10th| 933|
 11th| 1175|
 12th| 433|
 1st-4th| 168|
 5th-6th| 333|
 7th-8th| 646|
 9th| 514|
 Assoc-acdm| 1067|
 Assoc-voc| 1382|
 Bachelors| 5355|
 Doctorate| 413|
 HS-grad|10501|
 Masters| 1723|
 Preschool| 51|
 Prof-school| 576|
Some-college| 7291|
+------------+-----+

In [39]:
c = 'income'
adult_df \
  .groupBy(c) \
  .count() \
  .sort(col(c)).show()

+------+-----+
income|count|
+------+-----+
 <=50K|24720|
 >50K| 7841|
+------+-----+

In [40]:
c = 'maritalstatus'
adult_df \
  .groupBy(c) \
  .count() \
  .sort(col(c)).show()

+--------------------+-----+
 maritalstatus|count|
+--------------------+-----+
 Divorced| 4443|
 Married-AF-spouse| 23|
 Married-civ-spouse|14976|
Married-spouse-ab...| 418|
 Never-married|10683|
 Separated| 1025|
 Widowed| 993|
+--------------------+-----+

In [41]:
c = 'nativecountry'
adult_df \
  .groupBy(c) \
  .count() \
  .sort(col(c)).show()

+------------------+-----+
 nativecountry|count|
+------------------+-----+
 ?| 583|
 Cambodia| 19|
 Canada| 121|
 China| 75|
 Columbia| 59|
 Cuba| 95|
Dominican-Republic| 70|
 Ecuador| 28|
 El-Salvador| 106|
 England| 90|
 France| 29|
 Germany| 137|
 Greece| 29|
 Guatemala| 64|
 Haiti| 44|
Holand-Netherlands| 1|
 Honduras| 13|
 Hong| 20|
 Hungary| 13|
 India| 100|
+------------------+-----+
only showing top 20 rows

In [42]:
c = 'occupation'
adult_df \
  .groupBy(c) \
  .count() \
  .sort(col(c)).show()

+-----------------+-----+
 occupation|count|
+-----------------+-----+
 ?| 1843|
 Adm-clerical| 3770|
 Armed-Forces| 9|
 Craft-repair| 4099|
 Exec-managerial| 4066|
 Farming-fishing| 994|
Handlers-cleaners| 1370|
Machine-op-inspct| 2002|
 Other-service| 3295|
 Priv-house-serv| 149|
 Prof-specialty| 4140|
 Protective-serv| 649|
 Sales| 3650|
 Tech-support| 928|
 Transport-moving| 1597|
+-----------------+-----+

In [43]:
c = 'race'
adult_df \
  .groupBy(c) \
  .count() \
  .sort(col(c)).show()

+------------------+-----+
 race|count|
+------------------+-----+
Amer-Indian-Eskimo| 311|
Asian-Pac-Islander| 1039|
 Black| 3124|
 Other| 271|
 White|27816|
+------------------+-----+

In [44]:
c = 'relationship'
adult_df \
  .groupBy(c) \
  .count() \
  .sort(col(c)).show()

+--------------+-----+
 relationship|count|
+--------------+-----+
 Husband|13193|
 Not-in-family| 8305|
Other-relative| 981|
 Own-child| 5068|
 Unmarried| 3446|
 Wife| 1568|
+--------------+-----+

In [45]:
c = 'sex'
adult_df \
  .groupBy(c) \
  .count() \
  .sort(col(c)).show()

+------+-----+
 sex|count|
+------+-----+
Female|10771|
 Male|21790|
+------+-----+

In [46]:
c = 'workclass'
adult_df \
  .groupBy(c) \
  .count() \
  .sort(col(c)).show()

+----------------+-----+
 workclass|count|
+----------------+-----+
 ?| 1836|
 Federal-gov| 960|
 Local-gov| 2093|
 Never-worked| 7|
 Private|22696|
 Self-emp-inc| 1116|
Self-emp-not-inc| 2541|
 State-gov| 1298|
 Without-pay| 14|
+----------------+-----+

We can see that most of the values are acceptable.  However nativecountry, occupation, workclass all have values of "?" that are obviously illegitimate.

Now that we know where the issues are, let's talk about how to resolve them.  Strings are easiest in these scenarios.  The simplest way to deal with string values is simply to replace all of the illegitimate values with a single "Unknown" value.  In our case, these values are already using a standardized "?" value, so there's really no reason for us to change this.  Since this is a training exercise, let's go ahead and change them to "Unknown" using a loop.

In [49]:
from pyspark.sql.functions import when

cols = ['nativecountry', 'occupation', 'workclass']

for c in cols:
  adult_df = adult_df \
    .withColumn( c + '_u' \
      ,when( col(c) == '?', 'Unknown') \
        .otherwise(col(c)) \
    ) \
    .drop(c)

In [50]:
adult_df.show()

+---+-----------+-----------+------------+------------+------+------------+------+--------------------+------------------+-------------+------+---------------+-----------------+----------------+
age|capitalgain|capitalloss| education|educationnum|fnlwgt|hoursperweek|income| maritalstatus| race| relationship| sex|nativecountry_u| occupation_u| workclass_u|
+---+-----------+-----------+------------+------------+------+------------+------+--------------------+------------------+-------------+------+---------------+-----------------+----------------+
 39| 2174| 0| Bachelors| 13| 77516| 40| <=50K| Never-married| White|Not-in-family| Male| United-States| Adm-clerical| State-gov|
 50| 0| 0| Bachelors| 13| 83311| 13| <=50K| Married-civ-spouse| White| Husband| Male| United-States| Exec-managerial|Self-emp-not-inc|
 38| 0| 0| HS-grad| 9|215646| 40| <=50K| Divorced| White|Not-in-family| Male| United-States|Handlers-cleaners| Private|
 53| 0| 0| 11th| 7|234721| 40| <=50K| Married-civ-spouse| Black| Husband| Male| United-States|Handlers-cleaners| Private|
 28| 0| 0| Bachelors| 13|338409| 40| <=50K| Married-civ-spouse| Black| Wife|Female| Cuba| Prof-specialty| Private|
 37| 0| 0| Masters| 14|284582| 40| <=50K| Married-civ-spouse| White| Wife|Female| United-States| Exec-managerial| Private|
 49| 0| 0| 9th| 5|160187| 16| <=50K|Married-spouse-ab...| Black|Not-in-family|Female| Jamaica| Other-service| Private|
 52| 0| 0| HS-grad| 9|209642| 45| >50K| Married-civ-spouse| White| Husband| Male| United-States| Exec-managerial|Self-emp-not-inc|
 31| 14084| 0| Masters| 14| 45781| 50| >50K| Never-married| White|Not-in-family|Female| United-States| Prof-specialty| Private|
 42| 5178| 0| Bachelors| 13|159449| 40| >50K| Married-civ-spouse| White| Husband| Male| United-States| Exec-managerial| Private|
 37| 0| 0|Some-college| 10|280464| 80| >50K| Married-civ-spouse| Black| Husband| Male| United-States| Exec-managerial| Private|
 30| 0| 0| Bachelors| 13|141297| 40| >50K| Married-civ-spouse|Asian-Pac-Islander| Husband| Male| India| Prof-specialty| State-gov|
 23| 0| 0| Bachelors| 13|122272| 30| <=50K| Never-married| White| Own-child|Female| United-States| Adm-clerical| Private|
 32| 0| 0| Assoc-acdm| 12|205019| 50| <=50K| Never-married| Black|Not-in-family| Male| United-States| Sales| Private|
 40| 0| 0| Assoc-voc| 11|121772| 40| >50K| Married-civ-spouse|Asian-Pac-Islander| Husband| Male| Unknown| Craft-repair| Private|
 34| 0| 0| 7th-8th| 4|245487| 45| <=50K| Married-civ-spouse|Amer-Indian-Eskimo| Husband| Male| Mexico| Transport-moving| Private|
 25| 0| 0| HS-grad| 9|176756| 35| <=50K| Never-married| White| Own-child| Male| United-States| Farming-fishing|Self-emp-not-inc|
 32| 0| 0| HS-grad| 9|186824| 40| <=50K| Never-married| White| Unmarried| Male| United-States|Machine-op-inspct| Private|
 38| 0| 0| 11th| 7| 28887| 50| <=50K| Married-civ-spouse| White| Husband| Male| United-States| Sales| Private|
 43| 0| 0| Masters| 14|292175| 45| >50K| Divorced| White| Unmarried|Female| United-States| Exec-managerial|Self-emp-not-inc|
+---+-----------+-----------+------------+------------+------+------------+------+--------------------+------------------+-------------+------+---------------+-----------------+----------------+
only showing top 20 rows

In [51]:
c = 'workclass_u'
adult_df \
  .groupBy(c) \
  .count() \
  .sort(col(c)) \
  .show()

+----------------+-----+
 workclass_u|count|
+----------------+-----+
 Federal-gov| 960|
 Local-gov| 2093|
 Never-worked| 7|
 Private|22696|
 Self-emp-inc| 1116|
Self-emp-not-inc| 2541|
 State-gov| 1298|
 Unknown| 1836|
 Without-pay| 14|
+----------------+-----+

We can see that this process wasn't overly complicated with the use of a loop.  It's important to note that loops in python are tab-sensitive.  Meaning that all of the lines after the start of a loop must be tabbed out one place.  As we can see, this rule goes out the window once we start using "\" as a line break.

Let's move on to dealing with capitalgain.  As we saw earlier, this field has illegitimate values of 99999.  Unlike strings, there's no "Unknown" value we can provide.  Instead, we need to provide some type of number.  Mean and Median are pretty common alternatives.  Since this field represents money, we'll use median to minimize the skew.

In [54]:
adult_df \
  .approxQuantile('capitalgain', [0.5], 0.25)

Out[31]: [0.0]

The approxQuantile() function will approximate any Quantile within our dataset.  As some of you may know, Median is simply another word for the 50th quantile.  The results of this were not planned and lead us down an interesting tangent.  Is the median legitimately 0 or is there something else going on here?  Let's run a quick test to find out.

In [56]:
c = 'capitalgain'
adult_df \
  .withColumn( c + '_>0' \
    ,when( col(c) > 0, '>0' ) \
      .otherwise( '0' ) \
  ) \
  .groupBy(c + '_>0') \
  .count() \
  .show()

+--------------+-----+
capitalgain_>0|count|
+--------------+-----+
 0|29849|
 >0| 2712|
+--------------+-----+

We see that the vast majority of values in this column are 0.  This is very good evidence that we can use 0 as our imputed value.  Let's do that.

In [58]:
c = 'capitalgain'
adult_df = adult_df \
  .withColumn( c + '_0' \
    ,when( col(c) == 99999, 0 ) \
      .otherwise( col(c) ) \
  ) \
  .drop(c)

In [59]:
c = 'capitalgain_0'
adult_df \
  .groupBy(c) \
  .count() \
  .sort(col(c).desc()) \
  .show()

+-------------+-----+
capitalgain_0|count|
+-------------+-----+
 41310| 2|
 34095| 5|
 27828| 34|
 25236| 11|
 25124| 4|
 22040| 1|
 20051| 37|
 18481| 2|
 15831| 6|
 15024| 347|
 15020| 5|
 14344| 26|
 14084| 41|
 13550| 27|
 11678| 2|
 10605| 12|
 10566| 6|
 10520| 43|
 9562| 4|
 9386| 22|
+-------------+-----+
only showing top 20 rows

Now, there are no more values of 99999 in the capitalgain column.

Next, we need to prepare our text columns for modeling.  Most machine learning algorithms will only accept numeric columns.  A common technique for resolving this is to create indicator or dummy variables.  Let's start with the 'sex' column, as it only has a couple distinct values.

The first step is to create an array of the unique values in the column.

In [63]:
c = 'sex'
adult_df \
  .select(c) \
  .distinct() \
  .show()

+------+
 sex|
+------+
Female|
 Male|
+------+

Creating a DataFrame using the distinct() function is easy.  Now, we need to turn it into an array.

In [65]:
c = 'sex'
vals = adult_df \
  .select(c) \
  .distinct() \
  .rdd \
  .flatMap(lambda x: x) \
#.collect()

vals.collect()

Out[36]: ['Female', 'Male']

In [66]:
c = 'sex'
vals = adult_df \
  .select(c) \
  .distinct() \
  .rdd \
  .flatMap(lambda x: x) \
  .collect()

vals

Out[37]: ['Female', 'Male']

We can accomplish this by using the rdd() function to turn the DataFrame into an RDD containing Row objects.  Then, we can use the flatMap() function to turn this RDD of Row objects into an RDD of strings.  Finally, using the collect() function turns the RDD of strings into an array of strings that we can iterate over easily.

Now, we need to create new indicator columns for each unique value in the 'sex' column.

In [69]:
c = 'sex'
cols = \
  [ \
    when(col(c) == val, 1) \
    .otherwise(0)
    .alias( c + '_' + str(val).lower() ) \
    for val in vals \
  ]

cols

Out[38]: [Column<b'CASE WHEN (sex = Female) THEN 1 ELSE 0 END AS `sex_female`'>,
 Column<b'CASE WHEN (sex = Male) THEN 1 ELSE 0 END AS `sex_male`'>]

Here we see a loop within an array.  This allows us to create a foreach loop that loops through each element in an array, outputting the results to another array.  In this case, we are creating a column creation statement for each unique value within the 'sex' column.  Notice that this is a **COLUMN CREATION STATEMENT**, not the column itself.  This is an interesting Spark construct.  Basically, we can programmatically define the column creation as an object, then apply that column creation statement to any DataFrame we'd like.

Finally, we need to add these columns to our DataFrame and drop the original one.

In [72]:
c = 'sex'
adult_df \
    .select(adult_df.columns + cols) \
    .drop(c).show(5)

+---+-----------+---------+------------+------+------------+------+------------------+-----+-------------+---------------+-----------------+----------------+-------------+----------+--------+
age|capitalloss|education|educationnum|fnlwgt|hoursperweek|income| maritalstatus| race| relationship|nativecountry_u| occupation_u| workclass_u|capitalgain_0|sex_female|sex_male|
+---+-----------+---------+------------+------+------------+------+------------------+-----+-------------+---------------+-----------------+----------------+-------------+----------+--------+
 39| 0|Bachelors| 13| 77516| 40| <=50K| Never-married|White|Not-in-family| United-States| Adm-clerical| State-gov| 2174| 0| 1|
 50| 0|Bachelors| 13| 83311| 13| <=50K|Married-civ-spouse|White| Husband| United-States| Exec-managerial|Self-emp-not-inc| 0| 0| 1|
 38| 0| HS-grad| 9|215646| 40| <=50K| Divorced|White|Not-in-family| United-States|Handlers-cleaners| Private| 0| 0| 1|
 53| 0| 11th| 7|234721| 40| <=50K|Married-civ-spouse|Black| Husband| United-States|Handlers-cleaners| Private| 0| 0| 1|
 28| 0|Bachelors| 13|338409| 40| <=50K|Married-civ-spouse|Black| Wife| Cuba| Prof-specialty| Private| 0| 1| 0|
+---+-----------+---------+------------+------+------------+------+------------------+-----+-------------+---------------+-----------------+----------------+-------------+----------+--------+
only showing top 5 rows

While most of this isn't new, we do see a new use for the select() function.  Instead of supply string values separated by commas, we can supply a single array.  While doing this, we can make use of the + operator to concatenate the original set of columns with the new columns, thereby creating an array that contains all of the columns we want.

Now that we've seen all of the logic, let's wrap it in a loop to iterate through all of the string columns.

The first step is to create an array with all of the string columns.

In [76]:
strings = \
  [ \
    c[0] \
    for c in adult_df.dtypes \
    if c[1].startswith('string') \
  ]

strings

Out[40]: ['education',
 'income',
 'maritalstatus',
 'race',
 'relationship',
 'sex',
 'nativecountry_u',
 'occupation_u',
 'workclass_u']

Here, we see the loop within an array construct again.  In this case, we are looping through all of the values in the adult_df.dtypes array.  This array contains all of the names and data types for each column in the adult_df DataFrame.  We add some additional logic to this to store the first element (c[0], the column names) only if the second element (c[1], the column data types) starts with the letters 'string'.

Finally, let's use this array to loop through the entire DataFrame, creating all of the necessary dummy columns.

In [79]:
for c in strings:
  vals = \
    adult_df \
    .select(c) \
    .distinct() \
    .rdd \
    .flatMap(lambda x: x) \
    .collect()
  
  cols = \
    [ \
      when(col(c) == val, 1) \
     .otherwise(0)
     .alias( c + '_' + str(val).lower() ) \
     for val in vals \
    ]

  adult_df = adult_df \
    .select(adult_df.columns + cols) \
    .drop(c)
  
adult_df.columns

Out[41]: ['age',
 'capitalloss',
 'educationnum',
 'fnlwgt',
 'hoursperweek',
 'capitalgain_0',
 'education_masters',
 'education_10th',
 'education_5th-6th',
 'education_assoc-acdm',
 'education_assoc-voc',
 'education_7th-8th',
 'education_9th',
 'education_hs-grad',
 'education_bachelors',
 'education_11th',
 'education_1st-4th',
 'education_preschool',
 'education_12th',
 'education_doctorate',
 'education_some-college',
 'education_prof-school',
 'income_<=50k',
 'income_>50k',
 'maritalstatus_separated',
 'maritalstatus_never-married',
 'maritalstatus_married-spouse-absent',
 'maritalstatus_divorced',
 'maritalstatus_widowed',
 'maritalstatus_married-af-spouse',
 'maritalstatus_married-civ-spouse',
 'race_other',
 'race_amer-indian-eskimo',
 'race_white',
 'race_asian-pac-islander',
 'race_black',
 'relationship_own-child',
 'relationship_not-in-family',
 'relationship_unmarried',
 'relationship_wife',
 'relationship_other-relative',
 'relationship_husband',
 'sex_female',
 'sex_male',
 'nativecountry_u_philippines',
 'nativecountry_u_germany',
 'nativecountry_u_cambodia',
 'nativecountry_u_france',
 'nativecountry_u_greece',
 'nativecountry_u_taiwan',
 'nativecountry_u_ecuador',
 'nativecountry_u_nicaragua',
 'nativecountry_u_hong',
 'nativecountry_u_peru',
 'nativecountry_u_india',
 'nativecountry_u_china',
 'nativecountry_u_unknown',
 'nativecountry_u_italy',
 'nativecountry_u_holand-netherlands',
 'nativecountry_u_cuba',
 'nativecountry_u_south',
 'nativecountry_u_iran',
 'nativecountry_u_ireland',
 'nativecountry_u_thailand',
 'nativecountry_u_laos',
 'nativecountry_u_el-salvador',
 'nativecountry_u_mexico',
 'nativecountry_u_guatemala',
 'nativecountry_u_honduras',
 'nativecountry_u_yugoslavia',
 'nativecountry_u_puerto-rico',
 'nativecountry_u_jamaica',
 'nativecountry_u_canada',
 'nativecountry_u_united-states',
 'nativecountry_u_dominican-republic',
 'nativecountry_u_outlying-us(guam-usvi-etc)',
 'nativecountry_u_japan',
 'nativecountry_u_england',
 'nativecountry_u_haiti',
 'nativecountry_u_poland',
 'nativecountry_u_portugal',
 'nativecountry_u_columbia',
 'nativecountry_u_scotland',
 'nativecountry_u_hungary',
 'nativecountry_u_vietnam',
 'nativecountry_u_trinadad&tobago',
 'occupation_u_sales',
 'occupation_u_exec-managerial',
 'occupation_u_prof-specialty',
 'occupation_u_handlers-cleaners',
 'occupation_u_farming-fishing',
 'occupation_u_craft-repair',
 'occupation_u_unknown',
 'occupation_u_transport-moving',
 'occupation_u_priv-house-serv',
 'occupation_u_protective-serv',
 'occupation_u_other-service',
 'occupation_u_tech-support',
 'occupation_u_machine-op-inspct',
 'occupation_u_armed-forces',
 'occupation_u_adm-clerical',
 'workclass_u_self-emp-not-inc',
 'workclass_u_local-gov',
 'workclass_u_state-gov',
 'workclass_u_private',
 'workclass_u_unknown',
 'workclass_u_without-pay',
 'workclass_u_federal-gov',
 'workclass_u_never-worked',
 'workclass_u_self-emp-inc']

Now that we've created all of our new columns, there's one thing we need to take care of.  Mathematically speaking, we don't need all of these columns.  Since we've already dealt with any missing values, we know that each record in the dataset had a value for each original column.  Therefore, we also know that for each set of dummy variables, e.g. race\_x, exactly one of these dummy variables will be 1 and the others will be 0.  Therefore, we only need to keep (n - 1) of the columns.  For example, if we threw away the 'race\_other' column and saw that the other race columns were 0, then we know that the 'race\_other' column must have been 1.  The predictive models understand this too.  If we wanted to, could add some logic to our dummy variable creation process to remove the unnecessary columns.  However, they aren't hurting anything.  So, we'll leave them in for now.  But, there is one set of columns that we do need to deal with, income.  That's the column that we are ultimately trying to predict and we wouldn't want to predict 'income\_>50k' using the value 'income\_<=50k' because we would be correct 100% of the time.  So, we need to remove 'income\_<=50k' from the dataset.

In [81]:
adult_df = adult_df \
  .drop('income_<=50k')

adult_df.columns

Out[42]: ['age',
 'capitalloss',
 'educationnum',
 'fnlwgt',
 'hoursperweek',
 'capitalgain_0',
 'education_masters',
 'education_10th',
 'education_5th-6th',
 'education_assoc-acdm',
 'education_assoc-voc',
 'education_7th-8th',
 'education_9th',
 'education_hs-grad',
 'education_bachelors',
 'education_11th',
 'education_1st-4th',
 'education_preschool',
 'education_12th',
 'education_doctorate',
 'education_some-college',
 'education_prof-school',
 'income_>50k',
 'maritalstatus_separated',
 'maritalstatus_never-married',
 'maritalstatus_married-spouse-absent',
 'maritalstatus_divorced',
 'maritalstatus_widowed',
 'maritalstatus_married-af-spouse',
 'maritalstatus_married-civ-spouse',
 'race_other',
 'race_amer-indian-eskimo',
 'race_white',
 'race_asian-pac-islander',
 'race_black',
 'relationship_own-child',
 'relationship_not-in-family',
 'relationship_unmarried',
 'relationship_wife',
 'relationship_other-relative',
 'relationship_husband',
 'sex_female',
 'sex_male',
 'nativecountry_u_philippines',
 'nativecountry_u_germany',
 'nativecountry_u_cambodia',
 'nativecountry_u_france',
 'nativecountry_u_greece',
 'nativecountry_u_taiwan',
 'nativecountry_u_ecuador',
 'nativecountry_u_nicaragua',
 'nativecountry_u_hong',
 'nativecountry_u_peru',
 'nativecountry_u_india',
 'nativecountry_u_china',
 'nativecountry_u_unknown',
 'nativecountry_u_italy',
 'nativecountry_u_holand-netherlands',
 'nativecountry_u_cuba',
 'nativecountry_u_south',
 'nativecountry_u_iran',
 'nativecountry_u_ireland',
 'nativecountry_u_thailand',
 'nativecountry_u_laos',
 'nativecountry_u_el-salvador',
 'nativecountry_u_mexico',
 'nativecountry_u_guatemala',
 'nativecountry_u_honduras',
 'nativecountry_u_yugoslavia',
 'nativecountry_u_puerto-rico',
 'nativecountry_u_jamaica',
 'nativecountry_u_canada',
 'nativecountry_u_united-states',
 'nativecountry_u_dominican-republic',
 'nativecountry_u_outlying-us(guam-usvi-etc)',
 'nativecountry_u_japan',
 'nativecountry_u_england',
 'nativecountry_u_haiti',
 'nativecountry_u_poland',
 'nativecountry_u_portugal',
 'nativecountry_u_columbia',
 'nativecountry_u_scotland',
 'nativecountry_u_hungary',
 'nativecountry_u_vietnam',
 'nativecountry_u_trinadad&tobago',
 'occupation_u_sales',
 'occupation_u_exec-managerial',
 'occupation_u_prof-specialty',
 'occupation_u_handlers-cleaners',
 'occupation_u_farming-fishing',
 'occupation_u_craft-repair',
 'occupation_u_unknown',
 'occupation_u_transport-moving',
 'occupation_u_priv-house-serv',
 'occupation_u_protective-serv',
 'occupation_u_other-service',
 'occupation_u_tech-support',
 'occupation_u_machine-op-inspct',
 'occupation_u_armed-forces',
 'occupation_u_adm-clerical',
 'workclass_u_self-emp-not-inc',
 'workclass_u_local-gov',
 'workclass_u_state-gov',
 'workclass_u_private',
 'workclass_u_unknown',
 'workclass_u_without-pay',
 'workclass_u_federal-gov',
 'workclass_u_never-worked',
 'workclass_u_self-emp-inc']

Now, there are a few names that contain special characters.  While this might not have any impact, it's good practice to eliminate them whenever possible.  Specifically, these characters are -, >, (, ) and &.

We can start by creating an array of all the column names that need to be changed.

In [84]:
special = \
  [ \
    c \
    for c in adult_df.columns \
    if \
      '-' in c or \
      '>' in c or \
      '(' in c or \
      ')' in c or \
      '&' in c \
  ]

special

Out[43]: ['education_5th-6th',
 'education_assoc-acdm',
 'education_assoc-voc',
 'education_7th-8th',
 'education_hs-grad',
 'education_1st-4th',
 'education_some-college',
 'education_prof-school',
 'income_>50k',
 'maritalstatus_never-married',
 'maritalstatus_married-spouse-absent',
 'maritalstatus_married-af-spouse',
 'maritalstatus_married-civ-spouse',
 'race_amer-indian-eskimo',
 'race_asian-pac-islander',
 'relationship_own-child',
 'relationship_not-in-family',
 'relationship_other-relative',
 'nativecountry_u_holand-netherlands',
 'nativecountry_u_el-salvador',
 'nativecountry_u_puerto-rico',
 'nativecountry_u_united-states',
 'nativecountry_u_dominican-republic',
 'nativecountry_u_outlying-us(guam-usvi-etc)',
 'nativecountry_u_trinadad&tobago',
 'occupation_u_exec-managerial',
 'occupation_u_prof-specialty',
 'occupation_u_handlers-cleaners',
 'occupation_u_farming-fishing',
 'occupation_u_craft-repair',
 'occupation_u_transport-moving',
 'occupation_u_priv-house-serv',
 'occupation_u_protective-serv',
 'occupation_u_other-service',
 'occupation_u_tech-support',
 'occupation_u_machine-op-inspct',
 'occupation_u_armed-forces',
 'occupation_u_adm-clerical',
 'workclass_u_self-emp-not-inc',
 'workclass_u_local-gov',
 'workclass_u_state-gov',
 'workclass_u_without-pay',
 'workclass_u_federal-gov',
 'workclass_u_never-worked',
 'workclass_u_self-emp-inc']

We can see that this is relatively simple using the loop within an array construct again.  The only new functionality here is using the 'in' operator to search for strings within strings.

Now, let's replace all of the special characters in these column names.  We want to replace '<' with 'lt', '>' with 'gt', '=' with 'e' and everything else with '_'.

In [87]:
newcols = [ \
  c \
    .replace('-', '_') \
    .replace('>', 'gt') \
    .replace('(', '_') \
    .replace(')', '_') \
    .replace('&', '_') \
  for c in special
]

newcols

Out[44]: ['education_5th_6th',
 'education_assoc_acdm',
 'education_assoc_voc',
 'education_7th_8th',
 'education_hs_grad',
 'education_1st_4th',
 'education_some_college',
 'education_prof_school',
 'income_gt50k',
 'maritalstatus_never_married',
 'maritalstatus_married_spouse_absent',
 'maritalstatus_married_af_spouse',
 'maritalstatus_married_civ_spouse',
 'race_amer_indian_eskimo',
 'race_asian_pac_islander',
 'relationship_own_child',
 'relationship_not_in_family',
 'relationship_other_relative',
 'nativecountry_u_holand_netherlands',
 'nativecountry_u_el_salvador',
 'nativecountry_u_puerto_rico',
 'nativecountry_u_united_states',
 'nativecountry_u_dominican_republic',
 'nativecountry_u_outlying_us_guam_usvi_etc_',
 'nativecountry_u_trinadad_tobago',
 'occupation_u_exec_managerial',
 'occupation_u_prof_specialty',
 'occupation_u_handlers_cleaners',
 'occupation_u_farming_fishing',
 'occupation_u_craft_repair',
 'occupation_u_transport_moving',
 'occupation_u_priv_house_serv',
 'occupation_u_protective_serv',
 'occupation_u_other_service',
 'occupation_u_tech_support',
 'occupation_u_machine_op_inspct',
 'occupation_u_armed_forces',
 'occupation_u_adm_clerical',
 'workclass_u_self_emp_not_inc',
 'workclass_u_local_gov',
 'workclass_u_state_gov',
 'workclass_u_without_pay',
 'workclass_u_federal_gov',
 'workclass_u_never_worked',
 'workclass_u_self_emp_inc']

Again, we see the power of loops within arrays.  Here, we are using the .replace() function to substitute characters within our column names.

Now, let's rename these columns in our DataFrame.

In [90]:
for idx in range(0, len(special)):
  adult_df = adult_df \
    .withColumnRenamed(special[idx], newcols[idx])

adult_df.columns

Out[45]: ['age',
 'capitalloss',
 'educationnum',
 'fnlwgt',
 'hoursperweek',
 'capitalgain_0',
 'education_masters',
 'education_10th',
 'education_5th_6th',
 'education_assoc_acdm',
 'education_assoc_voc',
 'education_7th_8th',
 'education_9th',
 'education_hs_grad',
 'education_bachelors',
 'education_11th',
 'education_1st_4th',
 'education_preschool',
 'education_12th',
 'education_doctorate',
 'education_some_college',
 'education_prof_school',
 'income_gt50k',
 'maritalstatus_separated',
 'maritalstatus_never_married',
 'maritalstatus_married_spouse_absent',
 'maritalstatus_divorced',
 'maritalstatus_widowed',
 'maritalstatus_married_af_spouse',
 'maritalstatus_married_civ_spouse',
 'race_other',
 'race_amer_indian_eskimo',
 'race_white',
 'race_asian_pac_islander',
 'race_black',
 'relationship_own_child',
 'relationship_not_in_family',
 'relationship_unmarried',
 'relationship_wife',
 'relationship_other_relative',
 'relationship_husband',
 'sex_female',
 'sex_male',
 'nativecountry_u_philippines',
 'nativecountry_u_germany',
 'nativecountry_u_cambodia',
 'nativecountry_u_france',
 'nativecountry_u_greece',
 'nativecountry_u_taiwan',
 'nativecountry_u_ecuador',
 'nativecountry_u_nicaragua',
 'nativecountry_u_hong',
 'nativecountry_u_peru',
 'nativecountry_u_india',
 'nativecountry_u_china',
 'nativecountry_u_unknown',
 'nativecountry_u_italy',
 'nativecountry_u_holand_netherlands',
 'nativecountry_u_cuba',
 'nativecountry_u_south',
 'nativecountry_u_iran',
 'nativecountry_u_ireland',
 'nativecountry_u_thailand',
 'nativecountry_u_laos',
 'nativecountry_u_el_salvador',
 'nativecountry_u_mexico',
 'nativecountry_u_guatemala',
 'nativecountry_u_honduras',
 'nativecountry_u_yugoslavia',
 'nativecountry_u_puerto_rico',
 'nativecountry_u_jamaica',
 'nativecountry_u_canada',
 'nativecountry_u_united_states',
 'nativecountry_u_dominican_republic',
 'nativecountry_u_outlying_us_guam_usvi_etc_',
 'nativecountry_u_japan',
 'nativecountry_u_england',
 'nativecountry_u_haiti',
 'nativecountry_u_poland',
 'nativecountry_u_portugal',
 'nativecountry_u_columbia',
 'nativecountry_u_scotland',
 'nativecountry_u_hungary',
 'nativecountry_u_vietnam',
 'nativecountry_u_trinadad_tobago',
 'occupation_u_sales',
 'occupation_u_exec_managerial',
 'occupation_u_prof_specialty',
 'occupation_u_handlers_cleaners',
 'occupation_u_farming_fishing',
 'occupation_u_craft_repair',
 'occupation_u_unknown',
 'occupation_u_transport_moving',
 'occupation_u_priv_house_serv',
 'occupation_u_protective_serv',
 'occupation_u_other_service',
 'occupation_u_tech_support',
 'occupation_u_machine_op_inspct',
 'occupation_u_armed_forces',
 'occupation_u_adm_clerical',
 'workclass_u_self_emp_not_inc',
 'workclass_u_local_gov',
 'workclass_u_state_gov',
 'workclass_u_private',
 'workclass_u_unknown',
 'workclass_u_without_pay',
 'workclass_u_federal_gov',
 'workclass_u_never_worked',
 'workclass_u_self_emp_inc']

The only new functionality here is using the withColumnRenamed() function to rename columns in DataFrame.  Unfortunately, there was no way to pass the arrays directly into this function.  It would be possible using RDDs, but that would have been more complex.  Perhaps that's a good homework assignment if you want to tinker with it.

##Section 3: Data Science Modeling

Unlike many other Predictive Modeling tools, PySpark won't allow more than one Feature column.  This means that we have to combine all of our features into a single column before we can train our models.  This technique is called vectorization.  Let's give it a shot.

In [94]:
from pyspark.ml.feature import VectorAssembler

adult_vec = \
  VectorAssembler( \
    inputCols=[ \
      c \
      for c in adult_df.columns \
      if c != 'income_gt50k' \
    ] \
    ,outputCol="features" \
  ) \
  .transform(adult_df) \
  .select(['income_gt50k', 'features'])

adult_vec.head(5)

Out[46]: [Row(income_gt50k=0, features=SparseVector(108, {0: 39.0, 2: 13.0, 3: 77516.0, 4: 40.0, 5: 2174.0, 14: 1.0, 23: 1.0, 31: 1.0, 35: 1.0, 41: 1.0, 71: 1.0, 98: 1.0, 101: 1.0})),
 Row(income_gt50k=0, features=SparseVector(108, {0: 50.0, 2: 13.0, 3: 83311.0, 4: 13.0, 14: 1.0, 28: 1.0, 31: 1.0, 39: 1.0, 41: 1.0, 71: 1.0, 85: 1.0, 99: 1.0})),
 Row(income_gt50k=0, features=SparseVector(108, {0: 38.0, 2: 9.0, 3: 215646.0, 4: 40.0, 13: 1.0, 25: 1.0, 31: 1.0, 35: 1.0, 41: 1.0, 71: 1.0, 87: 1.0, 102: 1.0})),
 Row(income_gt50k=0, features=SparseVector(108, {0: 53.0, 2: 7.0, 3: 234721.0, 4: 40.0, 15: 1.0, 28: 1.0, 33: 1.0, 39: 1.0, 41: 1.0, 71: 1.0, 87: 1.0, 102: 1.0})),
 Row(income_gt50k=0, features=SparseVector(108, {0: 28.0, 2: 13.0, 3: 338409.0, 4: 40.0, 14: 1.0, 28: 1.0, 33: 1.0, 37: 1.0, 40: 1.0, 57: 1.0, 86: 1.0, 102: 1.0}))]

The VectorAssembler() function does most of the work for us here.  All we had to do was give it an array of the column names to vectorize, obviously leaving out the column that we are looking to predict.  This function also supports the .transform() function which can be used to apply it back to our dataset.

Viewing the resulting DataFrame as an RDD shows us the interesting structure of the Vector.  Basically, the SparseVector() contains a value with the total number of columns, as well as a key-value pair for each non-zero column.

Now that we our DataFrame has been appropriately transformed, we need to create our training testing splits.  This is very important so we can accurately evaluate whether our model is good or not.

In [97]:
(adult_train, adult_test) = adult_vec.randomSplit([0.7, 0.3])

{
  'Train': {
    'Rows': adult_train.count()
    ,'Cols': len(adult_train.columns)
  }
  ,'Test': {
    'Rows': adult_test.count()
    ,'Cols': len(adult_test.columns)
  }
}

Out[47]: {'Train': {'Rows': 22673, 'Cols': 2}, 'Test': {'Rows': 9888, 'Cols': 2}}

In this snippet, the .randomSplit() function is doing the hard work.  We do see a new convention in that we can retrieve two outputs from a single function by using parentheses.  Also, since Databricks will not output more than one result per cell, we need to create a dictionary to display all of our results at once.

Now that we're finally ready to start predictive modeling, let's train a Boosted Decision Tree using the default hyperparameters.

In [100]:
from pyspark.ml.classification import GBTClassifier

gbt = \
  GBTClassifier(featuresCol='features', labelCol='income_gt50k') \
  .fit(adult_train)

gbt

Out[48]: GBTClassificationModel (uid=GBTClassifier_3813cc55beb1) with 20 trees

The GBTClassifier() function creates an untrained model when we pass in the names of the label and features columns.  We can then apply the .fit() function to train this model against our training data.

Next, let's use this trained model to generate some predictions.

In [103]:
gbt_pred = gbt \
  .transform(adult_test)

gbt_pred.head(5)

Out[49]: [Row(income_gt50k=0, features=SparseVector(108, {0: 26.0, 1: 1876.0, 2: 14.0, 3: 397317.0, 4: 40.0, 6: 1.0, 23: 1.0, 31: 1.0, 35: 1.0, 40: 1.0, 71: 1.0, 86: 1.0, 102: 1.0}), rawPrediction=DenseVector([0.9208, -0.9208]), probability=DenseVector([0.8631, 0.1369]), prediction=0.0),
 Row(income_gt50k=0, features=SparseVector(108, {0: 40.0, 1: 1876.0, 2: 14.0, 3: 105717.0, 4: 35.0, 6: 1.0, 23: 1.0, 31: 1.0, 35: 1.0, 40: 1.0, 71: 1.0, 98: 1.0, 100: 1.0}), rawPrediction=DenseVector([0.6277, -0.6277]), probability=DenseVector([0.7782, 0.2218]), prediction=0.0),
 Row(income_gt50k=0, features=SparseVector(108, {0: 39.0, 1: 1974.0, 2: 14.0, 3: 49020.0, 4: 40.0, 6: 1.0, 23: 1.0, 31: 1.0, 35: 1.0, 41: 1.0, 71: 1.0, 85: 1.0, 102: 1.0}), rawPrediction=DenseVector([0.1753, -0.1753]), probability=DenseVector([0.5868, 0.4132]), prediction=0.0),
 Row(income_gt50k=0, features=SparseVector(108, {0: 44.0, 1: 1876.0, 2: 14.0, 3: 101593.0, 4: 42.0, 6: 1.0, 23: 1.0, 31: 1.0, 35: 1.0, 41: 1.0, 71: 1.0, 86: 1.0, 100: 1.0}), rawPrediction=DenseVector([0.4257, -0.4257]), probability=DenseVector([0.7008, 0.2992]), prediction=0.0),
 Row(income_gt50k=0, features=SparseVector(108, {0: 36.0, 1: 1408.0, 2: 14.0, 3: 255191.0, 4: 40.0, 6: 1.0, 23: 1.0, 31: 1.0, 35: 1.0, 41: 1.0, 71: 1.0, 86: 1.0, 105: 1.0}), rawPrediction=DenseVector([0.6972, -0.6972]), probability=DenseVector([0.8013, 0.1987]), prediction=0.0)]

We see that the trained model exposes the .transform() function for creating predictions.  This results in a few additional columns.  The rawPrediction column is a vector that show how likely each result is.  Since there were only two options for income_gt50k, there are only two values here.  The probability column is basically the same; however, it's standardized to the 0-1 scale that we are used to thinking about probabilities.  The probability column is simply the result with the highest probability.

Next, let's compare the predictions to the testing set to create some of the standard evaluation metrics.

In [106]:
from pyspark.ml.evaluation import BinaryClassificationEvaluator

evaluator = BinaryClassificationEvaluator(rawPredictionCol="rawPrediction", labelCol='income_gt50k')
gbt_auc = evaluator.evaluate(gbt_pred, {evaluator.metricName: "areaUnderROC"})
gbt_aupr = evaluator.evaluate(gbt_pred, {evaluator.metricName: "areaUnderPR"})

{
  'Area Under ROC Curve': gbt_auc
  ,'Area Under Precision-Recall Curve': gbt_aupr
}

Out[50]: {'Area Under ROC Curve': 0.9081299554984508,
 'Area Under Precision-Recall Curve': 0.7795283664944689}

Using the BinaryClassificationEvaluator() and .evaluate() functions, we see that our Boosted Decision Tree model has an AUC of .91 and an Area Under Precision-Recall Curve of .77.  These are pretty good results, especially for not doing any substantial feature engineering or model tuning.

We can use very similar code to create Logistic Regression and Naive Bayes models as well.  Let's check them out.

In [109]:
# Logistic Regression

from pyspark.ml.classification import LogisticRegression

lr = LogisticRegression(featuresCol='features', labelCol='income_gt50k') \
  .fit(adult_train)

lr_pred = lr \
  .transform(adult_test)

lr_auc = evaluator.evaluate(lr_pred, {evaluator.metricName: "areaUnderROC"})
lr_aupr = evaluator.evaluate(lr_pred, {evaluator.metricName: "areaUnderPR"})

# Naive Bayes

from pyspark.ml.classification import NaiveBayes

nb = NaiveBayes(featuresCol='features', labelCol='income_gt50k') \
  .fit(adult_train)

nb_pred = nb \
  .transform(adult_test)

nb_auc = evaluator.evaluate(nb_pred, {evaluator.metricName: "areaUnderROC"})
nb_aupr = evaluator.evaluate(nb_pred, {evaluator.metricName: "areaUnderPR"})

# Output Results

{
  'Area Under ROC Curve': {
    'Boosted Decision Tree': gbt_auc
    ,'Logistic Regression': lr_auc
    ,'Naive Bayes': nb_auc
  }
  ,'Area Under Precision-Recall Curve': {
    'Boosted Decision Tree': gbt_aupr
    ,'Logistic Regression': lr_aupr
    ,'Naive Bayes': nb_aupr
  }
}

Out[51]: {'Area Under ROC Curve': {'Boosted Decision Tree': 0.9081299554984508,
 'Logistic Regression': 0.9010054541262066,
 'Naive Bayes': 0.37380527997027163},
 'Area Under Precision-Recall Curve': {'Boosted Decision Tree': 0.7795283664944689,
 'Logistic Regression': 0.7588756122162874,
 'Naive Bayes': 0.19036180700971297}}

We see that the Logistic Regression model scored very closely to the Boosted Decision Tree, while the Naive Bayes model was substantially worse.

We hope you learned a lot by working through this Notebook.  We encourage you to continue to build on this knowledge and see what other amazing things you can create.

In [112]:
%sql

SELECT 1

1 1